src/middleware/etaghasher.rs

```diff
@@ -0,0 +1,318 @@
+pub trait RequestFilter<S> {
+    /// Return `true` if ETag processing should be applied to this
+    /// `(request, response)` pair and `false` otherwise. A `false` return
+    /// value will immediately return the original response unchanged.
+    fn filter(&self, req: &HttpRequest<S>, res: &HttpResponse) -> bool;
+}
+
+// Closure implementations
+impl<F: Fn(&[u8]) -> String> Hasher for F {
+    fn hash(&self, input: &[u8]) -> String {
+        self(input)
+    }
+}
+impl<S, F: Fn(&HttpRequest<S>, &HttpResponse) -> bool> RequestFilter<S> for F {
+    fn filter(&self, req: &HttpRequest<S>, res: &HttpResponse) -> bool {
+        self(req, res)
+    }
+}
+
+// Defaults
+/// Computes an ETag value from a byte slice using a default cryptographic hash
+/// function.
+pub struct DefaultHasher;
+impl Hasher for DefaultHasher {
+    fn hash(&self, input: &[u8]) -> String {
+        use sha1;
+        let mut h = sha1::Sha1::new();
+        h.update(input);
+        h.digest().to_string()
+    }
+}
+
+/// Returns `true` when the request method is `GET` or `HEAD` and the
+/// original response status is `200 OK`, and `false` otherwise.
+pub struct DefaultFilter;
+impl<S> RequestFilter<S> for DefaultFilter {
+    fn filter(&self, req: &HttpRequest<S>, res: &HttpResponse) -> bool {
+        use http::{Method, StatusCode};
+        (*req.method() == Method::GET || *req.method() == Method::HEAD)
+            && res.status() == StatusCode::OK
+    }
+}
+
+/// The middleware struct. Contains a Hasher to compute ETag values for byte
+/// slices and a filter to determine whether ETag computation and checking
+/// should be applied to a particular (request, response) pair.
+pub struct EtagHasher<S, H, F>
+where
+    S: 'static,
+    H: Hasher + 'static,
+    F: RequestFilter<S> + 'static,
+{
+    hasher: H,
+    filter: F,
+    _phantom: PhantomData<S>,
+}
+
+impl<S, H, F> EtagHasher<S, H, F>
+where
+    S: 'static,
+    H: Hasher + 'static,
+    F: RequestFilter<S> + 'static,
+{
+    /// Create a new middleware struct with the given Hasher and RequestFilter.
+    pub fn new(hasher: H, filter: F) -> Self {
+        EtagHasher {
+            hasher,
+            filter,
+            _phantom: PhantomData,
+        }
+    }
+}
+
+impl<S, H, F> middleware::Middleware<S> for EtagHasher<S, H, F>
+where
+    S: 'static,
+    H: Hasher + 'static,
+    F: RequestFilter<S> + 'static,
+{
+    fn response(
+        &mut self, req: &mut HttpRequest<S>, mut res: HttpResponse,
+    ) -> Result<middleware::Response> {
+        use header;
+        use Body;
+
+        if !self.filter.filter(req, &res) {
+            return Ok(middleware::Response::Done(res));
+        }
+
+        let e = if let Body::Binary(b) = res.body() {
+            Some(EntityTag::strong(self.hasher.hash(b.as_ref())))
+        } else {
+            None
+        };
+
+        if let Some(etag) = e {
+            if !none_match(&etag, req) {
```

Comment:
it seems like there is no need to return Option into `e`, just move this If section above where you check for `Body::Binary`?

Reply:
The code is structured this way so that the borrow taken when getting the body is dropped before we mutate the response.

Reply:
Ah I see, ok